### DISCUSSION TITLE
Conversions

### DISCUSSION THREAD

**Discussion by davecramer** [pgjdbc/r2dbc-postgresql#573] [General]
@mp911de 

Does r2dbc do conversions from types such as float4 to long ?

If so does it handle 10223372036850000000 properly ? This float is too large for a long.

**Comment by mp911de** [comment]
Guess you're asking about a forced cast instead of letting the driver determine the default type.

We first parse the value according to the database type

https://github.com/pgjdbc/r2dbc-postgresql/blob/87a058ec52006e83746ecb07e85b82b972b35402/src/main/java/io/r2dbc/postgresql/codec/NumericDecodeUtils.java#L46

and then we return a casted variant via `Number.floatValue()` (same pattern applies for integer and friends):

https://github.com/pgjdbc/r2dbc-postgresql/blob/87a058ec52006e83746ecb07e85b82b972b35402/src/main/java/io/r2dbc/postgresql/codec/FloatCodec.java#L39

The downcasting to smaller types comes with precision loss and we do not throw any exceptions in such a case.

**Comment by davecramer** [comment]
In the case of 10223372036850000000  it doesn't fit into a long ?

**Reply by mp911de** [reply]
That one would only fit into `NUMERIC` and we default to the `BigDecimal` return type (`new BigDecimal(String)` in text format, for binary decode we use a bit more complex code)
